custom_metrics/a11y.js

```diff
@@ -210,4 +210,184 @@ return JSON.stringify({
   screen_reader_classes: captureAndLogError(() => {
     return document.querySelectorAll('.sr-only, .visually-hidden').length > 0;
   }),
+  form_control_a11y_tree: captureAndLogError(() => {
+    function doesMatchAny(string, regexps) {
+      let found = false;
+      for (const regexp of regexps) {
+        if (regexp.test(string)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    const attributes_to_track = [
+      /^aria-.+$/,
+      /^type$/,
+      /^id$/,
+      /^name$/,
+      /^placeholder$/,
+      /^accept$/,
+      /^autocomplete$/,
+      /^autofocus$/,
+      /^capture$/,
+      /^max$/,
+      /^maxlength$/,
+      /^min$/,
+      /^minlength$/,
+      /^required$/,
+      /^readonly$/,
+      /^pattern$/,
+      /^multiple$/,
+      /^step$/,
+    ];
+    function addControlToStats(node, accumulator) {
+      const control_stats = {
+        type: node.node_info.nodeType.toLowerCase(),
+        attributes: {},
+        properties: {},
+        accessible_name: node.name.value || '',
+        accessible_name_sources: [],
+        role: node.role.value || '',
+      };
+
+      // Store all attribute information
+      for (let [key, value] of Object.entries(node.node_info.attributes || {})) {
+        key = key.toLowerCase();
+        if (!doesMatchAny(key, attributes_to_track)) {
+          continue;
+        }
+
+        control_stats.attributes[key] = value;
+      }
+
+      for (let property of node.properties) {
+        control_stats.properties[property.name] = property.value.value;
+      }
+
+      for (let source of node.name.sources || []) {
+        // Only include sources that contributed a value
+        if (!source.value || !source.value.value) {
+          continue;
+        }
+
+        // Only keep the relevant properties
+        const cleaned_source = {
+          type: source.type,
+          value: source.value.value,
+        };
+        if (source.attribute) {
+          cleaned_source.attribute = source.attribute;
+        }
+
+        control_stats.accessible_name_sources.push(cleaned_source);
+      }
+
+      accumulator.push(control_stats);
+    }
+
+    const allowed_control_types = [
+      'input',
+      'select',
+      'textarea',
+      'button',
+    ];
+
+    const stats_of_controls = [];
+    for (const node of $WPT_ACCESSIBILITY_TREE) {
```

Comment:
I'd suggest declaring a const at the top of the file that aliases `$WPT_ACCESSIBILITY_TREE`, so that there's only one version of it that needs to be expanded in case anyone wants to add a new custom metric that depends on it.

Reply:
Change made in latest commit